File: haskell/space-age/spaceAge.hs
module SpaceAge where

data Planet = Mercury 
            | Venus 
            | Earth 
            | Mars 
            | Jupiter 
            | Saturn 
            | Neptune 
            | Uranus
            
ageOn :: Planet -> Int -> Float 
ageOn planet secs = fromIntegral secs / (yearsPerOrbit * secsPerYear) 
        where secsPerYear = 60 * 60 * 24 * 365.25
              yearsPerOrbit = case planet of
                    Mercury -> 0.2408467  
                    Venus   -> 0.61519726 
                    Earth   -> 1.0
                    Mars    -> 1.8808158 
                    Jupiter -> 11.862615 
                    Saturn  -> 29.447498 
                    Uranus  -> 84.016846 
                    Neptune -> 164.79132 

File: ruby/space-age/space_age.rb
class SpaceAge
  SECS_IN_EARTH_YEAR = 31557600.0

  RELATIVE_YEARS = {
    "mercury" =>   0.2408467,
    "venus"   =>   0.61519726,
    "earth"   =>   1.0,
    "mars"    =>   1.8808158,
    "jupiter" =>  11.862615,
    "saturn"  =>  29.447498,
    "uranus"  =>  84.016846,
    "neptune" => 164.79132
  }

  attr_reader :seconds

  def initialize(age_in_secs)
    @seconds = age_in_secs.to_f
  end

  def method_missing(name, *args)
    return super unless name.match(/^on_(\w+)$/)
    if RELATIVE_YEARS.key?($1)
      age = @seconds / (RELATIVE_YEARS[$1] * SECS_IN_EARTH_YEAR)
      age.round(2)
    else
      return super
    end
  end
end

